### ISSUE TITLE
[Feature request] More text-shaping functionality.

### ISSUE THREAD

**Issue by LucidSigma** [libsdl-org/SDL_ttf#308]
## Font language.
This library currently has support for setting a font's script (via `TTF_SetFontScriptName`) and text direction (via `TTF_SetFontDirection`) to assist with text-shaping using HarfBuzz. However, HarfBuzz has another shaping setting — which is language. This is done with [hb_buffer_set_language](https://harfbuzz.github.io/harfbuzz-hb-buffer.html#hb-buffer-set-language).

A text's language is separate from its script. For example: Arabic, Persian, Urdu, and Sindhi all use an Arabic script, but have different alphabets and shaping rules.

I suggest the following API (to be congruent with the aforementioned functions):
```c
/*
    font: the font to specify the language for.
    language_bcp47: a null-terminated string containing the desired language's BCP47 code.
    returns: 0 on success, -1 on error.
*/
int TTF_SetFontLanguage(TTF_Font *font, const char *language_bcp47);
```

## Font beginning/end flags.
Some more advanced shaping features can be set using [hb_buffer_set_flags](https://harfbuzz.github.io/harfbuzz-hb-buffer.html#hb-buffer-set-flags). The list of available flags can be found [here](https://harfbuzz.github.io/harfbuzz-hb-buffer.html#hb-buffer-flags-t). As mentioned, most of these are advanced and out-of-scope for SDL_ttf; however I do feel that `HB_BUFFER_FLAG_BOT` and `HB_BUFFER_FLAG_EOT` could be implemented in SDL_ttf. These flags basically say that the shaping buffer's text is at the beginning/end of a paragraph (as some languages — such as Arabic — have specific text-shaping rules in these contexts).

Similarly to the previous suggestion, here is what the example API could look like herefor:
```c
/*
    font: the font to specify the beginning flag for.
    is_beginning: 0 to specify not at beginning of paragraph, non-zero otherwise.
*/
void TTF_SetFontBeginningOfParagraph(TTF_Font *font, int is_beginning);

/*
    font: the font to specify the end flag for.
    is_end: 0 to specify not at end of paragraph, non-zero otherwise.
*/
void TTF_SetFontEndOfParagraph(TTF_Font *font, int is_end);
```

**Comment by 1bsyl** [comment]
If this does better shaping, then yes, we should provide this.
I can create a PR

**Comment by 1bsyl** [comment]
Using the correct language (in addition to script), change a little bit the glyph/position for some language (and it's said to be orthogonal to script).

_"Languages are crucial for selecting which OpenType feature to apply to the buffer which can result in applying language-specific behaviour. Languages are orthogonal to the scripts, and though they are related, they are different concepts and should not be confused with each other."_


but, I saw no change testing the end/start of paragraph flags

**Comment by LucidSigma** [comment]
Awesome! Thanks for doing that.

I did a bit more research into the beginning/end flags. It turns out that the only real use for them (that I could find in the HarfBuzz source) was if a [combining character](https://en.wikipedia.org/wiki/Combining_character) was used at the start of the buffer. It also depended on the `HB_BUFFER_FLAG_DO_NOT_INSERT_DOTTED_CIRCLE` flag as well.

After seeing this, I feel that supporting them would be a bit too niche for SDL_ttf (especially if it adds new symbols to the API). If someone really needs to set the buffer flags, then they're better off just using HarfBuzz directly.

**Comment by 1bsyl** [comment]
ok, that make sense. if you cut the buffer, in the middle of a cluster (because of fixed buffer size maybe), the character cannot be rendered.

please give a try to the SetLanguage API, it sounds it should be added.

**Comment by LucidSigma** [comment]
If you think these flags should be supported, then I'll implement them.

I think we should support all of the buffer flags that HarfBuzz provides, since some are coupled with each other (for example: `HB_BUFFER_FLAG_BOT` and `HB_BUFFER_FLAG_BOT`).

I suggest the following API for them:
```c
typedef enum {
    TTF_SHAPING_FLAG_DEFAULT                        = 0x00000000,
    TTF_SHAPING_FLAG_BEGINNING_OF_TEXT              = 0x00000001,
    TTF_SHAPING_FLAG_END_OF_TEXT                    = 0x00000002,
    TTF_SHAPING_FLAG_PRESERVE_DEFAULT_IGNORABLES    = 0x00000004,
    TTF_SHAPING_FLAG_REMOVE_DEFAULT_IGNORABLES      = 0x00000008,
    TTF_SHAPING_FLAG_DO_NOT_INSERT_DOTTED_CIRCLE    = 0x00000010,
    TTF_SHAPING_FLAG_VERIFY                         = 0x00000020,
    TTF_SHAPING_FLAG_PRODUCE_UNSAFE_TO_CONCAT       = 0x00000040,
    TTF_SHAPING_FLAG_PRODUCE_SAFE_TO_INSERT_TATWEEL = 0x00000080,
    TTF_SHAPING_FLAG_ALL                            = 0x000000FF,
} TTF_ShapingFlags;

extern DECLSPEC int SDLCALL TTF_SetFontShapingFlags(TTF_Font *font, TTF_ShapingFlags flags);
```

The `TTF_ShapingFlags` enumeration basically contains all of the values from [`hb_buffer_flags_t`](https://harfbuzz.github.io/harfbuzz-hb-buffer.html#hb-buffer-flags-t).
The `TTF_SetFontShapingFlags` function sets the buffer flags for the respective font.

I'll await your feedback for this and then I implement it and create a pull request.

**Comment by 1bsyl** [comment]
I don´t  think we really need them. 
the ttf api allows text strings with any length, so that you can always render the string you want with the api.
either, you cannot really cut the string and render in two parts, because you would have two sdl_surface in the end.

but maybe I am wrong and not really familiar with language that would produce a wrong output.
 (if you have some counter-example ...)

**Comment by LucidSigma** [comment]
I definitely agree. Most of these flags are for advanced/multi-step shaping — and I feel if anyone really needs them, then they're probably better off making their own font renderer instead of using SDL_ttf.

Unless there's anything else, I'd say that we can close this issue once your PR is merged.
